src/ui/AttachResult/AttachResult.ts

```diff
@@ -0,0 +1,180 @@
+                    resultUriHash: this.queryResult.raw.urihash,
+                    author: this.queryResult.raw.author,
+                    articleID: null,
+                    caseID: null
+                };
+          
+                this.usageAnalytics.logCustomEvent<IAnalyticsCaseAttachMeta>(analyticsActionCauseList.caseAttach, customData, this.root);
+            })
+            .finally(() => { this.setLoading(false); });
+    }
+
+    /**
+     * Detach the query result.
+     */
+    public detach() : Promise<void> {
+        if (!this.attached && !this.loading) {
+            return Promise.resolve();
+        }
+
+        this.setLoading(true);
+        return this.options.detach(this.queryResult)
+            .then(() => { this.attached = false; })
+            .then(() => {
+                let customData: IAnalyticsCaseAttachMeta = {
+                    resultUriHash: this.queryResult.raw.urihash,
+                    author: this.queryResult.raw.author,
+                    articleID: null,
+                    caseID: null
+                };
+          
+                this.usageAnalytics.logCustomEvent<IAnalyticsCaseDetachMeta>(analyticsActionCauseList.caseDetach, customData, this.root);
+            })
+            .finally(() => { this.setLoading(false); });
+    }
+
+    /** Toggle the state of the component. If the current result is not attached, attach it, if not, detach it. */
+    public toggleAttached() : void {
+        if (this.attached) {
+            this.detach();
+        } else {
+            this.attach();
+        }
+    }
+
+    protected initialize() : void {
+        // Adds HTML elements needed to display this component.
+        this.buttonElement = $$("div", {}, PaperclipIcon).el;
+        this.element.appendChild(this.buttonElement);
+
+        this.tooltipElement = $$("div", { className: "coveo-caption-for-icon" }).el;
+        this.element.appendChild(this.tooltipElement);
+
+        this.attached = false;
+        this.render();
+
+        // Resolve the current result for the component and the initial state.
+        if (this.options.isAttached) {
+            this.setLoading(true);
+            this.options.isAttached(this.queryResult)
+                .then(attached => { this.attached = attached; })
+                .catch(error => { 
+                    this.logger.error("Error retrieving initial result attached state.", error);
+                })
+                .finally(() => { this.setLoading(false); });
+        }
+    }
+
+    /** Set the loading property and updates the component UI. */
+    protected setLoading(loading: boolean) : void {
+        this.loading = loading;
+        this.render();
+    }
+
+    protected render() : void {
+        $$(this.buttonElement).removeClass("coveo-icon-attached");
+        $$(this.buttonElement).removeClass("coveo-icon-attach");
+        $$(this.buttonElement).removeClass("coveo-icon-loading");
+
+        if (this.loading) {
```

Comment:
You could extract some of that logic into a private method: getClassFromState or something of the sort that would return either "coveo-icon-loading" or "coveo-icon-attached" or "coveo-icon-attach".

Reply:
it's already kind of small and i still have to add the text so not sure it's worth it